File: packages/mac/Sources/TruthPulseCore/Services/SearchService.swift
import Foundation

/// Pre-computed search index entry for a single market.
private struct IndexedMarket: Sendable {
    let market: MarketSummary
    let haystack: String // pre-normalized, joined searchable text
}

public actor SearchService {
    private let apiClient: KalshiAPIClient
    private let cacheStore: MarketCacheStore
    private let rankingPolicy = RankingPolicy()
    private let synonymTable = SynonymTable()
    private var trendCache: [String: MarketTrend] = [:]
    private var lastRefresh: Date?
    private var markets: [MarketSummary] = []
    private var index: [IndexedMarket] = []

    /// Synchronous check — no actor hop needed. Safe because it only does a file stat.
    nonisolated public let hasCacheOnDisk: Bool

    public init(apiClient: KalshiAPIClient = KalshiAPIClient()) throws {
        let appSupport = try SearchService.makeAppSupportDirectory()
        self.apiClient = apiClient
        self.cacheStore = MarketCacheStore(appSupportDirectory: appSupport)
        self.hasCacheOnDisk = cacheStore.cacheFileExists()
    }

    public func bootstrapIfNeeded() async throws {
        if markets.isEmpty {
            let loaded = await cacheStore.loadMarkets()
            setMarkets(loaded)
        }
    }

    public func hasLoadedMarkets() async -> Bool {
        if markets.isEmpty {
            let loaded = await cacheStore.loadMarkets()
            setMarkets(loaded)
        }
        return !markets.isEmpty
    }

    public func lastCacheDate() async -> Date? {
        await cacheStore.savedAt()
    }

    public func refreshOpenMarkets(force: Bool = false) async throws {
        if !force, let lastRefresh, Date().timeIntervalSince(lastRefresh) < 45 {
            return
        }

        let fetched = try await apiClient.fetchOpenMarkets()
        setMarkets(fetched)
        await cacheStore.saveMarkets(fetched)
        self.lastRefresh = Date()
    }

    private func setMarkets(_ newMarkets: [MarketSummary]) {
        markets = newMarkets
        index = newMarkets.map { market in
            let fields = [
                market.title,
                market.subtitle ?? "",
                market.yesLabel ?? "",
                market.noLabel ?? "",
                market.eventTitle ?? "",
                market.eventSubtitle ?? "",
                market.category ?? "",
                market.ticker
            ]
            let haystack = fields.joined(separator: " ").normalizedSearchText
            return IndexedMarket(market: market, haystack: haystack)
        }
    }

    /// All currently loaded markets (for Spotlight indexing).
    public var allMarkets: [MarketSummary] { markets }

    public func search(query: String, limit: Int = 30) -> [SearchResult] {
        let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return [] }

        let normalizedQuery = trimmed.normalizedSearchText
        let tokens = normalizedQuery.searchTokens
        let expandedTokens = synonymTable.expandTokens(tokens)

        let candidates = index
            .filter { entry in
                // Match on original query phrase
                if entry.haystack.contains(normalizedQuery) { return true }
                // Match if all original tokens present
                if tokens.allSatisfy({ entry.haystack.contains($0) }) { return true }
                // Match if all expanded tokens (with synonyms) present
                if expandedTokens.count > tokens.count {
                    return expandedTokens.allSatisfy { entry.haystack.contains($0) }
                }
                return false
            }
            .map { IndexedSearchCandidate(market: $0.market, baseRank: 0) }

        return Array(rankingPolicy.rank(query: query, candidates: candidates).prefix(limit))
    }

    public func trend(for marketTicker: String, seriesTicker: String, window: TrendWindow) async throws -> MarketTrend {
        let key = cacheKey(for: marketTicker, window: window)
        if let cached = trendCache[key], Date().timeIntervalSince(cached.updatedAt) < 300 {
            return cached
        }

        let trend = try await apiClient.fetchTrend(for: marketTicker, seriesTicker: seriesTicker, window: window)
        trendCache[key] = trend
        return trend
    }

    private func cacheKey(for marketTicker: String, window: TrendWindow) -> String {
        "\(marketTicker)::\(window.rawValue)"
    }

    public static func makeAppSupportDirectory() throws -> URL {
        let base = try FileManager.default.url(
            for: .applicationSupportDirectory,
            in: .userDomainMask,
            appropriateFor: nil,
            create: true
        )
        let directory = base.appendingPathComponent("TruthPulse", isDirectory: true)
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        return directory
    }

}

File: packages/android/app/src/main/java/com/truthpulse/search/SearchService.kt
package com.truthpulse.search

import android.content.Context
import com.truthpulse.data.*
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * Core search service: maintains market index, search, and trend caching.
 * Mirrors the Swift actor-based SearchService.
 */
class SearchService(context: Context) {

    private val apiClient = KalshiApiClient()
    private val cacheStore = MarketCacheStore(context)
    val synonymTable = SynonymTable(context)
    private val rankingPolicy = RankingPolicy()

    private val mutex = Mutex()
    private var markets: List<MarketSummary> = emptyList()
    private var index: List<IndexedMarket> = emptyList()
    private var lastRefreshMillis: Long? = null
    private val trendCache = mutableMapOf<String, MarketTrend>()

    val hasCacheOnDisk: Boolean get() = cacheStore.cacheFileExists()

    val marketCount: Int get() = markets.size

    suspend fun bootstrapIfNeeded() {
        mutex.withLock {
            if (markets.isEmpty()) {
                val loaded = cacheStore.loadMarkets()
                setMarketsInternal(loaded)
            }
        }
    }

    suspend fun hasLoadedMarkets(): Boolean {
        mutex.withLock {
            if (markets.isEmpty()) {
                val loaded = cacheStore.loadMarkets()
                setMarketsInternal(loaded)
            }
            return markets.isNotEmpty()
        }
    }

    fun lastCacheDateMillis(): Long? = cacheStore.savedAtMillis()

    /**
     * Refresh markets from the API with a 45-second debounce.
     */
    suspend fun refreshOpenMarkets(force: Boolean = false) {
        mutex.withLock {
            val now = System.currentTimeMillis()
            if (!force) {
                lastRefreshMillis?.let { last ->
                    if (now - last < 45_000) return
                }
            }
        }

        val fetched = apiClient.fetchOpenMarkets()

        mutex.withLock {
            setMarketsInternal(fetched)
            lastRefreshMillis = System.currentTimeMillis()
        }

        cacheStore.saveMarkets(fetched)
    }

    private fun setMarketsInternal(newMarkets: List<MarketSummary>) {
        markets = newMarkets
        index = newMarkets.map { market ->
            val fields = listOf(
                market.title,
                market.subtitle ?: "",
                market.yesLabel ?: "",
                market.noLabel ?: "",
                market.eventTitle ?: "",
                market.eventSubtitle ?: "",
                market.category ?: "",
                market.ticker,
            )
            val haystack = fields.joinToString(" ").normalizedSearchText()
            IndexedMarket(market = market, haystack = haystack)
        }
    }

    /**
     * Search markets using two-pass filtering with synonym expansion.
     */
    fun search(query: String, limit: Int = 30): List<SearchResult> {
        val trimmed = query.trim()
        if (trimmed.isEmpty()) return emptyList()

        val normalizedQuery = trimmed.normalizedSearchText()
        val tokens = normalizedQuery.searchTokens()
        val expandedTokens = synonymTable.expandTokens(tokens)

        val candidates = index
            .filter { entry ->
                // Pass 1: phrase match
                if (entry.haystack.contains(normalizedQuery)) return@filter true
                // Pass 2: all original tokens
                if (tokens.all { entry.haystack.contains(it) }) return@filter true
                // Pass 3: expanded tokens (synonym expansion)
                if (expandedTokens.size > tokens.size) {
                    return@filter expandedTokens.all { entry.haystack.contains(it) }
                }
                false
            }
            .map { IndexedSearchCandidate(market = it.market, baseRank = 0.0) }

        return rankingPolicy.rank(query, candidates).take(limit)
    }

    /**
     * Fetch trend data with 5-minute cache.
     */
    suspend fun trend(marketTicker: String, seriesTicker: String, window: TrendWindow): MarketTrend {
        val key = "$marketTicker::${window.name}"
        val now = System.currentTimeMillis()

        mutex.withLock {
            trendCache[key]?.let { cached ->
                if (now - cached.updatedAtMillis < 300_000) return cached
            }
        }

        val trend = apiClient.fetchTrend(marketTicker, seriesTicker, window)

        mutex.withLock {
            trendCache[key] = trend
        }

        return trend
    }

    /** All currently loaded markets (for AppSearch indexing). */
    fun allMarkets(): List<MarketSummary> = markets
}

private data class IndexedMarket(
    val market: MarketSummary,
    val haystack: String,
)
